### ISSUE TITLE
REQUEST-921-PROTOCOL-ATTACK

### ISSUE THREAD

**Issue by astew-msft** [coreruleset/coreruleset#3286]
Issue: WAF/Prevention Mode is not detecting CR/LF in the form of: GET /%0d%0aMy-Custom-Header:My-Custom-Value%0d%0a
![originalRequestUriWithArgs_s](https://github.com/coreruleset/coreruleset/assets/143033714/21cf77c8-8f1b-4afb-9449-35be9d3e0a86)

Expectation is that since these requests contain “/%0d%0a” the request should be blocked by the WAF. Request is only visible in ApplicationGatewayAccessLog but no corresponding Match/Block in the ApplicationGatewayFirewallLog.

Category: ApplicationGatewayAccessLog
requestUri_s: /
Set-Cookie:SRT=AnInsiderThreat
httpMethod_s: GET
WAFEvaluationTime_s: 0.000
WAFMode_s: Prevention
details_matches_s: -
details_msg_s: -
ruleSetType_s: -
ruleSetVersion_s: -
ruleId_s: -
action_s: -
transactionId_g: 94061d4e-ed9b-bdb8-c02c-687d36ab2183
originalRequestUriWithArgs_s: /%0d%0aSet-Cookie:SRT=AnInsiderThreat%0d%0a
serverStatus_s: 303

https://learn.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-crs-rulegroups-rules?tabs=owasp32#crs921-32

Does rule evaluate URI/URL with ARGs for CR/LF?

**Comment by dune73** [comment]
Hey @astew-msft, sorry for the inconvenience. But it seems this works on native CRS. Could it be an Azure problem?

```
$ curl -H "x-crs-paranoia-level: 1" -H "x-format-output: txt-matched-rules"  "https://sandbox.coreruleset.org/%0d%0aMy-Custom-Header:My-Custom-Value%0d%0a"
921190 PL1 HTTP Splitting (CR/LF in request filename detected)
...
```

If my call is wrong and I did not get your message, please let me know.

**Comment by astew-msft** [comment]
Hey Christian,


the OWASP rule set for rule 921151 applies to ARGS_GET.



If I put %0a in the ARGS then the request is blocked, if I put it outside of the request it is let through.

[image]


Is this by design, as In the OWASP rule set it only looks for CR/LF in the ARGS.



Does OWASP rule that blocks CR/LF in the URL/URI path.


***@***.***

***@***.***
Adrian Stewart
Azure Customer Engineer Gov

Phone: +1 470-826-8198
Email: ***@***.******@***.***>

Working Hours: Mon-Fri 8am-5pm EST
OOF:


***@***.******@***.***<https://www.credly.com/badges/10fbfdd6-5aa8-4363-bc27-b0d7a0c76584/public_url>

We are always interested to hear you feedback. Please feel free to email my manager/Myriam Johnson regarding the level of service you have received, at ***@***.******@***.***>

If you need immediate attention for any of the ongoing case while I am Out of Facility:

  *   email Gov escalations at ***@***.******@***.***>
  *   email Gov backup support at ***@***.******@***.***>

**Comment by dune73** [comment]
Sorry, but I do not follow you.

"If I put %0a in the ARGS then the request is blocked, if I put it outside of the request it is let through."

No, that's not the case. As I showed above, there is 921190 that checks the request filename. And there is also a header rule that does the same.

* 921140 PL1 : `REQUEST_HEADERS_NAMES|REQUEST_HEADERS "@rx [\n\r]"`
* 921150 PL1 : `ARGS_NAMES "@rx [\n\r]"`
* 921151 PL2 : `ARGS_GET "@rx [\n\r]"`
* 921160 PL1 : `ARGS_GET_NAMES|ARGS_GET "@rx [\n\r]+...` (list of certain HTTP headers)
* 921190 PL1 : `REQUEST_FILENAME "@rx [\n\r]"`

Where do you think the detection is lacking in this regard?
And then there is also 921160

**Comment by astew-msft** [comment]
Sorry for the inconvenience. Think detection is lacking URI. 

Mitigation: Create a custom rule with block action based on URI.

**Comment by astew-msft** [comment]
confirm CRS 3.2 is being used.





***@***.***

***@***.***
Adrian Stewart
Azure Customer Engineer Gov

Phone: +1 470-826-8198
Email: ***@***.******@***.***>

Working Hours: Mon-Fri 8am-5pm EST
OOF:


***@***.***<https://www.credly.com/badges/dfc17e6e-63e8-4618-8842-062a3e89ab55/public_url> ***@***.*** <https://www.credly.com/badges/10fbfdd6-5aa8-4363-bc27-b0d7a0c76584/public_url>

We are always interested to hear you feedback. Please feel free to email my manager/Myriam Johnson regarding the level of service you have received, at ***@***.******@***.***>

If you need immediate attention for any of the ongoing case while I am Out of Facility:

  *   email Gov escalations at ***@***.******@***.***>
  *   email Gov backup support at ***@***.******@***.***>

**Comment by theseion** [comment]
@astew-msft Considering that your issue is with 3.2 and that you had no further inquiries, I'm going to close this issue. Please reopen it if you need more help.
